maven/WFLY-14934_Server_Plugin.adoc

```diff
@@ -0,0 +1,155 @@
+
+* [ ] Managed domain
+
+* [x] OpenShift s2i
+
+* [x] Maven-controlled standalone server
+
+* [ ] Bootable jar
+
+== Requirements
+
+Ability to provision and configure a server installation.
+
+=== Hard Requirements
+
+==== Plugin requirements
+
+* Plugin produces inside the project target dir a directory (named server by default) that contains the server and any deployment.
+* By default the server is generated in ```target/server``` directory. The plugin offers a configuration item to provide the name of a directory relative to the project 
+   ```target``` directory in which the server will be installed. E.g.: ```<output-dir-name>foo-server</output-dir-name>```
+
+==== Plugin goals
+
+* ```wildfly-server:help```: Display plugin help
+* ```wildfly-server:package```: Build the application and server.
+
+==== Plugin Configuration items
+
+The detailed configuration can be found in the plugin https://github.com/jfdenise/wildfly-server-maven-plugin/wiki[community documentation]
+
+Plugin configuration items overview:
+
+* cli-sessions: List of CLI execution sessions
+* context-root: To make the WAR deployment registered under root resource path ('/').
+* display-cli-scripts-output: By default executed CLI scripts output is not shown if execution
+* excluded-layers: A list of Galleon layers to exclude
+* extra-server-content-dirs: A list of directories to copy content to the provisioned server
+* feature-pack-location: The WildFly Galleon feature-pack location
+* feature-packs: A list of feature-pack configurations to install
+* hollow-server: Create a server that doesn’t contain application
+* layers: A list of Galleon layers to provision.
+* log-time: Whether to log provisioning time at the end
+* off-line: Whether to use offline mode when the plugin resolves an artifact
+* output-dir-name: By default the generated server is 'server'
+* plugin-options: Arbitrary Galleon options used when provisioning the server
+* project-build-dir: Project build dir.
+* provisioning-file: The path to the provisioning.xml file to use
+* record-state: Whether to record provisioned state in .galleon directory.
+* skip: To skip execution
+
+==== Plugin artifacts
+
+===== WildFly Server Maven Plugin
+
+The maven plugin GAV is ```org.wildfly.plugins:wildfly-server-maven-plugin:1.0.0.Final```.
+The source repository is expected to be ```http://github.com/wildfly-extras/wildfly-server-plugin```
+
+===== WildFly Bootable JAR and WildFly Server plugins shared artifacts
+
+In order to avoid duplicating some logic common with WildFly Bootable JAR Maven plugin, the ```org.wildfly.plugins:wildfly-jar-maven-plugin-common:1.0.0.Final``` and ```org.wildfly.plugins:wildfly-jar-maven-tests-common:1.0.0.Final```
+Maven artifacts are defined. 
+Both WildFly Bootable JAR and WildFly Server Maven plugins depend on these artifacts.
+The source repository is expected to be ```http://github.com/wildfly-extras/wildfly-jar-maven-plugin-common```
```

Comment:
if we create a new common libray for `wildfly-server-maven-plugin` and `wildfly-jar-maven-plugin`, let's name it `wildfly-common-maven-plugin` instead of `wildfly-jar-maven-plugin-common`

Have we also considered having all these Maven modules inside a single GitHub repository:

```
+ http://github.com/wildfly-extras/wildfly-maven-plugins
   + common -> `org.wildfly.plugins:wildfly-comon-maven-plugin`
   + server -> `org.wildfly.plugins:wildfly-server-maven-plugin`
   + jar -> `org.wildfly.plugins:wildfly-jar-maven-plugin`
```

Reply:
In order to un-couple the life cycles of the plugins, I was thinking to separate the repositories and have each repo to have its own versioning and life cycle. Fixing a bug in bootable JAR shouldn't imply a bump in server plugin.
The bootable JAR is released for each new WildFly Major (mainly for examples but we have a bootable JAR runtime in WF that evolved sometime in sync with the plugin).
For the WildFly server plugin, I hope that we can uncouple the plugin life cycle from WF life cycle. We shouldn't release a new one for new WF Major. This plugin is very generic.

Reply:
ok